internal/controllers/topology/cluster/desired_state.go

```diff
@@ -626,7 +626,7 @@ func computeMachineDeployment(_ context.Context, s *scope.Scope, desiredControlP
 	// If it's a new MachineDeployment, set the finalizer.
 	// Note: we only add it on creation to avoid race conditions later on when
 	// the MachineDeployment topology controller removes the finalizer.
-	if currentMachineDeployment == nil {
+	if currentMachineDeployment == nil || currentMachineDeployment.Object.DeletionTimestamp.IsZero() {
 		controllerutil.AddFinalizer(desiredMachineDeploymentObj, clusterv1.MachineDeploymentTopologyFinalizer)
```

Comment:
I think we also have to add it if !currentMachineDeployment.Object.DeletionTimestamp.IsZero() && currentMachineDeployment still has the finalizer.

Otherwise the topology reconciler would remove the finalizer as soon as the deletionTimestamp is set, but this should only ever be done by the topology MD reconciler.

Reply:
Essentially we should
* add the finalizer if new Deployment and currentMachineDeployment.Object.DeletionTimestamp.IsZero()
* never drop the finalizer if it exists on currentMachineDeployment (and for that we have to add it here)